### PR TITLE
feat(ui): enable students to delete their rating tutors

### PR DESCRIPTION
**Pull request by mostafakamar2308**

### CODE REVIEW

File: packages/headless/src/rating.ts [hunk 1]
```diff
@@ -55,3 +55,26 @@ export function useEditRateTutor({
     mutationKey: ["edit-rate", id],
   });
 }
+
+export function useDeleteRateTutor({
+  id,
```
[the remark this answers is not in the document]
Comment (mostafakamar2308) [reply packages/headless/src/rating.ts:74]: In the custom hook itself, I am using it as a key for mutationKey

[the remark this answers is not in the document]
Comment (neuodev) [reply packages/headless/src/rating.ts:74]: It is fine to remove it.

File: apps/nova/src/components/TutorProfile/DeleteRating.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  Dialog,
+  toaster,
+  useFormatMessage,
+} from "@litespace/luna";
+import { useQueryClient } from "@tanstack/react-query";
+import { useDeleteRatingTutor } from "@litespace/headless/rating";
+
+type DeleteRatingProps = {
+  open: boolean;
+  close: () => void;
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/DeleteRating.tsx:14]:
```suggestion
  close: Void;
```

File: apps/nova/src/components/TutorProfile/DeleteRating.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  open,
+  close,
+}) => {
+  const queryClient = useQueryClient();
+  const intl = useFormatMessage();
+  const onSuccess = () => {
+    toaster.success({ title: intl("tutor.rate.delete.success") });
+    queryClient.invalidateQueries({ queryKey: ["tutor-rating", tutorId] });
+    close();
+  };
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/DeleteRating.tsx:31]:
```suggestion
  const invalidate = useInvalidateQuery();
  const intl = useFormatMessage();
  const onSuccess = () => {
    toaster.success({ title: intl("tutor.rate.delete.success") });
    invalidate(QueryKey.TutorRatings);
    close();
  };
```

File: apps/nova/src/components/TutorProfile/DeleteRating.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+const DeleteRating: React.FC<DeleteRatingProps> = ({
+  id,
+  tutorId,
+  open,
+  close,
+}) => {
+  const queryClient = useQueryClient();
+  const intl = useFormatMessage();
+  const onSuccess = () => {
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/DeleteRating.tsx:27]: Please wrap it with `useCallback`

File: apps/nova/src/components/TutorProfile/DeleteRating.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    toaster.success({ title: intl("tutor.rate.delete.success") });
+    queryClient.invalidateQueries({ queryKey: ["tutor-rating", tutorId] });
+    close();
+  };
+  const onError = (error: Error) => {
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/DeleteRating.tsx:32]: It is better to wrap it with `useCallback`

File: apps/nova/src/components/TutorProfile/DeleteRating.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    });
+    close();
+  };
+
+  const deleteRating = useDeleteRatingTutor({ onSuccess, onError, id });
+
+  const action = useMemo(() => {
+    return {
+      label: intl("global.labels.confirm"),
+      onClick: () => deleteRating.mutate(),
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/DeleteRating.tsx:45]:
```suggestion
      onClick: () => deleteRating.mutate(id),
```

File: apps/nova/src/components/TutorProfile/DeleteRating.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  };
+  const onError = (error: Error) => {
+    toaster.error({
+      title: intl("tutor.rate.delete.error"),
+      description: error.message,
+    });
+    close();
+  };
+
+  const deleteRating = useDeleteRatingTutor({ onSuccess, onError, id });
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/DeleteRating.tsx:40]:
```suggestion
  const deleteRating = useDeleteRatingTutor({ onSuccess, onError });
```

File: apps/nova/src/components/TutorProfile/RateForm.tsx [hunk 8]
```diff
@@ -28,27 +31,26 @@ type RateFormProps = {
 
 const RateForm: React.FC<RateFormProps> = ({ tutor, rate, close }) => {
   const intl = useFormatMessage();
-  const queryClient = useQueryClient();
+  const invalidateRatings = useInvalidateQuery();
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/RateForm.tsx:34]:
```suggestion
  const invalidate= useInvalidateQuery();
```

File: apps/nova/src/components/TutorProfile/Rating.tsx [hunk 9]
```diff
@@ -4,21 +4,24 @@ import {
...
 
 const Rating: React.FC<{ rating: IRating.Populated }> = ({ rating }) => {
+  const profile = useAppSelector(profileSelectors.user);
   const edit = useRender();
   const deleteRating = useRender();
-  const intl = useFormatMessage();
-  const showEditDialog = ()=> edit.show();
-  const showDeleteDialog = () => deleteRating.show()
+  const showEditDialog = () => edit.show();
+  const showDeleteDialog = () => deleteRating.show();
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Rating.tsx:23]:
```suggestion
```

File: apps/nova/src/components/TutorProfile/Rating.tsx [hunk 10]
```diff
@@ -4,21 +4,24 @@ import {
...
 import { useRender } from "@/hooks/render";
 import EditRating from "./EditRating";
+import { profileSelectors } from "@/redux/user/profile";
+import DeleteRating from "./DeleteRating";
+import { useAppSelector } from "@/redux/store";
 
 const Rating: React.FC<{ rating: IRating.Populated }> = ({ rating }) => {
+  const profile = useAppSelector(profileSelectors.user);
   const edit = useRender();
   const deleteRating = useRender();
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Rating.tsx:21]: For consistency, we can rename `edit` to `editRating`
```suggestion
  const editRating = useRender();
  const deleteRating = useRender();
```

File: apps/nova/src/components/TutorProfile/Rating.tsx [hunk 11]
```diff
@@ -46,22 +49,38 @@ const Rating: React.FC<{ rating: IRating.Populated }> = ({ rating }) => {
...
-              <Edit className="w-4 text-white" />
-            </Button>
-            <Button onClick={showDeleteDialog} size={ButtonSize.Tiny} type={ButtonType.Primary}>
-              <Trash2 className="w-4 text-white" />
-            </Button>
-          </div>
+          {profile?.id === rating.rater.id && (
+            <div className="flex gap-3">
+              <Button
+                onClick={showEditDialog}
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Rating.tsx:55]:
```suggestion
                onClick={showRating.show}
```

File: apps/nova/src/components/TutorProfile/Rating.tsx [hunk 12]
```diff
@@ -46,22 +49,38 @@ const Rating: React.FC<{ rating: IRating.Populated }> = ({ rating }) => {
...
+                size={ButtonSize.Tiny}
+                type={ButtonType.Primary}
+              >
+                <Edit className="w-4 text-white" />
+              </Button>
+              <Button
+                onClick={showDeleteDialog}
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Rating.tsx:62]:
```suggestion
                onClick={deleteRating.show}
```

File: apps/nova/src/components/TutorProfile/Rating.tsx [hunk 13]
```diff
@@ -46,22 +49,38 @@ const Rating: React.FC<{ rating: IRating.Populated }> = ({ rating }) => {
...
+            <EditRating open={edit.open} close={edit.hide} rate={rating} />
+
+            <DeleteRating
+              tutorId={rating.ratee.id}
+              open={deleteRating.open}
+              close={deleteRating.hide}
+              id={rating.id}
+            />
+          </>
+        )}
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Rating.tsx:83]: Let's use the ternary operator syntax when conditionally render components. Not to accidentally render something unexpected. 

```suggestion
        {profile?.id === rating.rater.id ? (
          <>
            <EditRating open={edit.open} close={edit.hide} rate={rating} />

            <DeleteRating
              tutorId={rating.ratee.id}
              open={deleteRating.open}
              close={deleteRating.hide}
              id={rating.id}
            />
          </>
        ) : null}
```

File: packages/headless/src/query.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import { useCallback } from "react";
+import { useQueryClient } from "@tanstack/react-query";
+
+export function useInvalidateQuery() {
+  const client = useQueryClient();
+  return useCallback(
+    (keys: (string | number)[]) => client.invalidateQueries({ queryKey: keys }),
```
Comment (neuodev) [inline packages/headless/src/query.ts:7]: Nice!

**Review by neuodev — CHANGES_REQUESTED** [review]

File: packages/headless/src/rating.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mostafakamar2308) [reply packages/headless/src/rating.ts:74]: Alright

**Comment by mostafakamar2308** [comment]
Change requests implemented 🚀

**Comment by neuodev** [comment]
> Change requests implemented 🚀

Great! 

Please click "Re-request" review once you done with the changes and want to request another review. 

image

File: apps/nova/src/components/TutorProfile/DeleteRating.tsx [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+  close,
+}) => {
+  const invalidate = useInvalidateQuery();
+  const intl = useFormatMessage();
+
+  const onSuccess = useCallback(() => {
+    toaster.success({ title: intl("tutor.rate.delete.success") });
+    invalidate([QueryKeys.FindTutorRating, tutorId]);
+    close();
+  }, []);
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/DeleteRating.tsx:34]: Please add empty line. 

```suggestion
  }, []);
  
```

File: apps/nova/src/components/TutorProfile/DeleteRating.tsx [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+  tutorId,
+  open,
+  close,
+}) => {
+  const invalidate = useInvalidateQuery();
+  const intl = useFormatMessage();
+
+  const onSuccess = useCallback(() => {
+    toaster.success({ title: intl("tutor.rate.delete.success") });
+    invalidate([QueryKeys.FindTutorRating, tutorId]);
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/DeleteRating.tsx:32]: Nice!

File: apps/nova/src/components/TutorProfile/DeleteRating.tsx [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+  const deleteRating = useDeleteRatingTutor({ onSuccess, onError });
+
+  const action = useMemo(() => {
+    return {
+      label: intl("global.labels.confirm"),
+      onClick: () => deleteRating.mutate(id),
+      disabled: deleteRating.isPending,
+      loading: deleteRating.isPending,
+    };
+  }, [intl, deleteRating]);
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/DeleteRating.tsx:52]: Please add empty line

```suggestion
  }, [intl, deleteRating]);
  
```

Comment (neuodev) [inline apps/nova/src/components/TutorProfile/DeleteRating.tsx:52]: Perfect!

File: apps/nova/src/components/TutorProfile/Rating.tsx [hunk 18]
```diff
@@ -63,15 +64,25 @@ const Rating: React.FC<{ rating: IRating.Populated }> = ({ rating }) => {
...
-      </div>
+      ) : null}
+      {profile?.id === rating.rater.id ? (
+        <DeleteRating
+          tutorId={rating.ratee.id}
+          open={deleteRating.open}
+          close={deleteRating.hide}
+          id={rating.id}
+        />
+      ) : null}
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Rating.tsx:85]: The problem with this approach that we render `DeleteRating` and `EditRating` for each rating component we render on the screen. 

A better approach will be to something like this:

```ts
// TutorProfile.tsx
const [editRating, setEditRating] = useState<IRate.Self | null>(null)
const [deleteRating, setDeleteRating] = useState<IRate.Self | null>(null)

return (
   <Stats query={stats} />
  <About about={tutor.data.about} />
  <ActivitiesOverview query={activity} />
  <Ratings query={ratings} setEditRating={setEditRating} setDeleteRating={setDeleteRating}  />
  {editRating && profile.id === editRating.rater.id ? <EditRating {...props}  /> : null}
  {deleteRating && profile.id === deleteRating.rater.id ? <DeleteRating {...props}  /> : null}
)
```

Using the approach mentioned above we only render one edit/delete dialog for the entire ratings list. this is more elegant, clean, performant. 

@mostafakamar2308 WDYT?

**Review by neuodev — CHANGES_REQUESTED** [review]

File: apps/nova/src/components/TutorProfile/Rating.tsx (hunk 18 above)
Comment (mostafakamar2308) [reply apps/nova/src/components/TutorProfile/Rating.tsx]: I agree with you on that, it's certainly more clean and elegant.
I will change it today, إن شاء الله. 🚀

**Comment by neuodev** [comment]
@mostafakamar2308 I merged #4, please rebase your branch on the latest master.

File: apps/nova/src/components/TutorProfile/RateForm.tsx [hunk 19]
```diff
@@ -28,26 +31,27 @@ type RateFormProps = {
...
+
   const onSuccess = useCallback(() => {
-    queryClient.invalidateQueries({ queryKey: ["tutor-rating", tutor] });
+    console.log(tutor);
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/RateForm.tsx:37]: Left over, please delete it. 
```suggestion
```

**Review by neuodev — APPROVED** [review]
Great work @mostafakamar2308! I left a minor comment. Please take a look

File: apps/nova/src/components/TutorProfile/RateForm.tsx [hunk 20]
```diff
@@ -28,26 +31,26 @@ type RateFormProps = {
 
 const RateForm: React.FC<RateFormProps> = ({ tutor, rate, close }) => {
   const intl = useFormatMessage();
-  const queryClient = useQueryClient();
+  const invalidateRatings = useInvalidateQuery();
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/RateForm.tsx:34]:
```suggestion
  const invalidate = useInvalidateQuery();
```

File: apps/nova/src/components/TutorProfile/Rating.tsx [hunk 21]
```diff
@@ -9,15 +9,22 @@ import { IRating } from "@litespace/types";
...
-const Rating: React.FC<{ rating: IRating.Populated }> = ({ rating }) => {
+  handleEdit: (rating: IRating.Populated) => void;
+  handleDelete: ({
+    ratingId,
+    tutorId,
+  }: {
+    ratingId: number;
+    tutorId: number;
+  }) => void;
+}> = ({ rating, handleEdit, handleDelete }) => {
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Rating.tsx:26]:
```suggestion
  onEdit: (rating: IRating.Populated) => void;
  onDelete: (payload: {
    ratingId: number;
    tutorId: number;
  }) => void;
}> = ({ rating, handleEdit, handleDelete }) => {
```

File: apps/nova/src/components/TutorProfile/Ratings.tsx [hunk 22]
```diff
@@ -1,22 +1,72 @@
...
+import EditRating from "@/components/TutorProfile/EditRating";
+import DeleteRating from "@/components/TutorProfile/DeleteRating";
 
 const Ratings: React.FC<{
   query: UseQueryResult<IRating.Populated[], Error>;
 }> = ({ query }) => {
+  const [ratingEdited, setRatedEdited] = useState<IRating.Populated | null>(
+    null
+  );
+  const [ratingDeleted, setRatingDeleted] = useState<{
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Ratings.tsx:18]:
```suggestion
  const [editRating, setEditRating] = useState<IRating.Populated | null>(
    null
  );
  const [deleteRating, setDeleteRating] = useState<{
```

File: apps/nova/src/components/TutorProfile/Ratings.tsx [hunk 23]
```diff
@@ -1,22 +1,72 @@
...
+    ratingId: number;
+    tutorId: number;
+  }>({
+    ratingId: 0,
+    tutorId: 0,
+  });
+
+  const editRating = useRender();
+  const deleteRating = useRender();
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Ratings.tsx:27]: We no longer need `useRender`  as we can check if `editRating` or `deleteRating` is not `null`

```suggestion
```

File: apps/nova/src/components/TutorProfile/Ratings.tsx [hunk 24]
```diff
@@ -1,22 +1,72 @@
...
+
+  const handleEdit = useCallback((rating: IRating.Populated) => {
+    editRating.show();
+    setRatedEdited(rating);
+  }, []);
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Ratings.tsx:32]:
```suggestion
  const onEditRating = useCallback((rating: IRating.Populated) => setEditRating(rating), []);
```

File: apps/nova/src/components/TutorProfile/Ratings.tsx [hunk 25]
```diff
@@ -1,22 +1,72 @@
...
+
+  const handleDelete = useCallback(
+    ({ ratingId, tutorId }: { ratingId: number; tutorId: number }) => {
+      deleteRating.show();
+      setRatingDeleted({ ratingId, tutorId });
+    },
+    []
+  );
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Ratings.tsx:40]:
```suggestion
  const onDeleteRating = useCallback(
    (payload: { ratingId: number; tutorId: number }) => setRatingDeleted(payload),
    []
  );
```

File: apps/nova/src/components/TutorProfile/Ratings.tsx [hunk 26]
```diff
@@ -1,22 +1,72 @@
...
+    setRatedEdited(null);
+    editRating.hide();
+  }, []);
+  const handleCloseDeleteRating = useCallback(() => {
+    setRatingDeleted({
+      ratingId: 0,
+      tutorId: 0,
+    });
+    deleteRating.hide();
+  }, []);
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Ratings.tsx:52]:
```suggestion
  const onClose = useCallback(() => 
    setEditRating(null);
    setDeleteRating(null);
  }, []);  
```

File: apps/nova/src/components/TutorProfile/Ratings.tsx [hunk 27]
```diff
@@ -1,22 +1,72 @@
...
+  const [ratingEdited, setRatedEdited] = useState<IRating.Populated | null>(
+    null
+  );
+  const [ratingDeleted, setRatingDeleted] = useState<{
+    ratingId: number;
+    tutorId: number;
+  }>({
+    ratingId: 0,
+    tutorId: 0,
+  });
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Ratings.tsx:24]: It is better to initialize it to null. 

```suggestion
    ratingId: number;
    tutorId: number;
  } | null>(null);
```

File: apps/nova/src/components/TutorProfile/Ratings.tsx [hunk 28]
```diff
@@ -25,6 +75,21 @@ const Ratings: React.FC<{
...
+        />
+      ) : null}
+      {ratingDeleted ? (
+        <DeleteRating
+          tutorId={ratingDeleted.tutorId}
+          open={deleteRating.open}
+          close={handleCloseDeleteRating}
+          id={ratingDeleted.ratingId}
+        />
+      ) : null}
```
Comment (neuodev) [inline apps/nova/src/components/TutorProfile/Ratings.tsx:92]:
```suggestion
      {editRating ? (
        <EditRating
          open
          close={onClose}
          rate={editRating}
        />
      ) : null}
      
      {deleteRating ? (
        <DeleteRating
          open
          tutorId={deleteRating.tutorId}
          close={onClose}
          id={deleteRating.ratingId}
        />
      ) : null}
```

**Review by neuodev — APPROVED** [review]

**Pull request merged**